src/contexts/ApiPromiseContextProvider.tsx

```diff
@@ -42,29 +42,48 @@ export function ApiPromiseContextProvider(props: ApiRxContextProviderProps): Rea
   const options = { hasher, provider, types };
   const [apiPromise, setApiPromise] = useState<ApiPromise>(new ApiPromise(options));
   const [isReady, setIsReady] = useState(false);
+  const [disconnected, setDisconnected] = useState(false);
 
   useEffect(() => {
     if (isReady) {
+      logger.info(`${contextType} was changed`);
       setIsReady(false);
-      apiPromise.disconnect().then(() => logger.info(`${contextType} Resetting connection`));
+      setApiPromise(new ApiPromise(options));
     }
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [sourceTarget]);
 
   useDidUpdateEffect(() => {
-    ApiPromise.create(options).then((_api) => {
-      setApiPromise(_api);
-    });
-  }, [provider]);
+    if (!isReady && !apiPromise.isConnected) {
+      apiPromise.disconnect().then(() => {
+        setDisconnected(true);
+        logger.info(`${contextType} Resetting connection`);
+      });
+    }
+  }, [isReady]);
 
-  useEffect(() => {
-    apiPromise.isReady.then(() => {
-      if (types) {
-        registry.register(types);
-      }
+  useDidUpdateEffect(() => {
+    if (!apiPromise.isConnected) {
+      ApiPromise.create(options).then((_api) => {
+        logger.info(`${contextType} connection recreated `);
+        setApiPromise(_api);
+        setDisconnected(false);
+      });
+    }
+  }, [apiPromise.isConnected]);
 
-      setIsReady(true);
-    });
+  useEffect(() => {
+    if (!disconnected) {
+      apiPromise.isReady
+        .then(() => {
+          if (types) {
+            registry.register(types);
+          }
+          logger.info(`${contextType} type registration ready`);
+          setIsReady(true);
+        })
+        .catch((e) => logger.error(`${contextType} error in registration`, e));
+    }
```

Comment:
TBH I have no idea what is happening in all of these `use*Effect` handlers and why we need so many, but if it's necessary - so be it.